Fix the bug where the localStorage state is invalid.

venkatesh920 opened the pull request.

venkatesh920 committed:
Fix a bug and minor refactor of photo.js

venkatesh920 committed:
fix one more thing

venkatesh920 committed:
Squashing latest two commits into one

venkatesh920 committed:
Merge branch 'master' of https://github.com/venkatesh920/Chrome-EP-NewTab

IngridMorstrad reviewed (changes requested):
Please provide a better commit message so the reviewer can identify what was fixed in each of the 2  commits.

IngridMorstrad commented on javascript/photo.js:
```diff
@@ -51,12 +51,12 @@ backgrounds.Photo = new Model({
-        var numbers = Array();
-        var ind, temp;
+        var i, ind, temp, numbers;
+        numbers = [];
```
[Question] What's the advantage of this over Array() ?

venkatesh920 replied:
No advantage.  They're the same object.  Just seemed like this was more conventional.  Either way is fine - I don't have a strong preference.

IngridMorstrad commented on javascript/photo.js:
```diff
@@ -115,64 +115,65 @@ backgrounds.Photo = new Model({
+        var count = localStorage.getItem("count");
+        var order = localStorage.getItem("order");
+        var cachedImages = localStorage.getItem("cachedImages");
+        return count != null && Number(count) >= 0 && order != null && JSON.parse(order).length == this.BACKUP_IMAGES_COUNT && cachedImages != null && JSON.parse(cachedImages).length <= cacheSize;
```
Is count >= 0 required? Just a sanity check? Similar comments about the other checks.
Also, if possible split this return into three boolean checks for readability if possible.

IngridMorstrad commented on javascript/photo.js:
```diff
@@ -115,64 +115,65 @@ backgrounds.Photo = new Model({
      * Displays the image.
-     * TODO: Update function name & doc. This function does more than just display
+     * TODO: Update function name & doc. This function does more than just
+     * display
```
Why was this split to 2 lines when the rest of the file exceeds this line length in many places?

IngridMorstrad commented on javascript/photo.js:
```diff
@@ -115,64 +115,65 @@ backgrounds.Photo = new Model({
     display: function(numberOfImages, cacheSize) {
-        var count = localStorage.getItem("count");
-        if (count === null) {
+        if (!this.isValidLocalStorageState(cacheSize)) {
```
Nice!

IngridMorstrad commented on javascript/photo.js:
```diff
@@ -115,64 +115,65 @@ backgrounds.Photo = new Model({
-                var cachedImages = JSON.parse(localStorage.getItem("cachedImages"));
-                //var cacheIndex = Number(localStorage.getItem("cacheIndex"));
-                cachedImages[cacheIndex] = imageData;
+                cachedImages.push(imageData);
```
Does this work even though cachedImages is not defined within this function? We had to use parent (this) earlier to solve this.

venkatesh920 replied:
cachedImages is in the same scope as parent is.  So it should be fine.

IngridMorstrad replied:
Ah... good way of explaining it!

IngridMorstrad commented on javascript/photo.js:
```diff
@@ -115,64 +115,65 @@ backgrounds.Photo = new Model({
-        // TODO: Think about not calling the API after we get all images (or caching all images & then not calling?)
-        xmlHttp.open("GET", "https://www.reddit.com/r/EarthPorn/top/.json?limit=" + Number(numberOfImages), true);
-        xmlHttp.send(null);
+        // We send an HTTP request only if our cache is not full.
```
When will it be equal? Assume cacheSize > 1, then each time we add to the cache, on the next call, we will pull from it.

venkatesh920 replied:
If the startup image count exceeds the cacheSize, then the cache will become full before we can use it.

IngridMorstrad commented on javascript/photo.js:
```diff
@@ -115,64 +115,65 @@ backgrounds.Photo = new Model({
-            this.displayImage(cachedImages[cachedImageNumber]);
+            this.displayImage(cachedImages[0]);
+            // Once we consume an image from the cache, remove it from the cache.
+            cachedImages.splice(0, 1);
```
When will the cache size be > this.STARTUP_IMAGE_COUNT? cacheSize variable seems unnecessary.

venkatesh920 replied:
STARTUP_IMAGE_COUNT and cacheSize shouldn't have any dependency on one another.  For example, STARTUP_IMAGE_COUNT could be equal to BACKUP_IMAGES_COUNT and that shouldn't break how we use the cache.

IngridMorstrad reviewed (changes requested):
Nice work overall... :)

venkatesh920 committed:
Nit: Fix comment on line 137.

The comment on line 137 was broken into two lines.  Bring it back to how it was (one line).

venkatesh920 committed:
Improve readability of isValidLocalStorageState

venkatesh920 commented:
BTW, we'll squash all the commits into one before merging so don't bother about the individual commits; rather just look at the final diff.
